File: extensions/music/src/main/java/app/revanced/extension/music/patches/NavigationBarPatch.java
package app.revanced.extension.music.patches;

import static app.revanced.extension.shared.Utils.hideViewUnderCondition;

import android.view.View;
import android.widget.TextView;

import androidx.annotation.Nullable;

import java.util.Arrays;
import java.util.List;

import app.revanced.extension.music.settings.Settings;

@SuppressWarnings("unused")
public class NavigationBarPatch {
    private static String lastYTNavigationEnumName = "";

    public static void setLastAppNavigationEnum(@Nullable Enum<?> ytNavigationEnumName) {
        if (ytNavigationEnumName != null) {
            lastYTNavigationEnumName = ytNavigationEnumName.name();
        }
    }

    public static void hideNavigationLabel(TextView textview) {
        hideViewUnderCondition(Settings.HIDE_NAVIGATION_BAR_LABEL.get(), textview);
    }

    public static void hideNavigationButton(View view) {
        // Hide entire navigation bar.
        if (Settings.HIDE_NAVIGATION_BAR.get() && view.getParent() != null) {
            hideViewUnderCondition(true, (View) view.getParent());
            return;
        }

        // Hide navigation buttons based on their type.
        for (NavigationButton button : NavigationButton.values()) {
            if (button.ytEnumNames.contains(lastYTNavigationEnumName)) {
                hideViewUnderCondition(button.hidden, view);
                break;
            }
        }
    }

    private enum NavigationButton {
        HOME(
                Arrays.asList(
                        "TAB_HOME"
                ),
                Settings.HIDE_NAVIGATION_BAR_HOME_BUTTON.get()
        ),
        SAMPLES(
                Arrays.asList(
                        "TAB_SAMPLES"
                ),
                Settings.HIDE_NAVIGATION_BAR_SAMPLES_BUTTON.get()
        ),
        EXPLORE(
                Arrays.asList(
                        "TAB_EXPLORE"
                ),
                Settings.HIDE_NAVIGATION_BAR_EXPLORE_BUTTON.get()
        ),
        LIBRARY(
                Arrays.asList(
                        "LIBRARY_MUSIC",
                        "TAB_BOOKMARK" // YouTube Music 8.24+
                ),
                Settings.HIDE_NAVIGATION_BAR_LIBRARY_BUTTON.get()
        ),
        UPGRADE(
                Arrays.asList(
                        "TAB_MUSIC_PREMIUM"
                ),
                Settings.HIDE_NAVIGATION_BAR_UPGRADE_BUTTON.get()
        );

        private final List<String> ytEnumNames;
        private final boolean hidden;

        NavigationButton(List<String> ytEnumNames, boolean hidden) {
            this.ytEnumNames = ytEnumNames;
            this.hidden = hidden;
        }
    }
}

File: patches/src/main/kotlin/app/revanced/patches/music/layout/navigationbar/NavigationBarPatch.kt
package app.revanced.patches.music.layout.navigationbar

import app.revanced.patcher.extensions.InstructionExtensions.addInstruction
import app.revanced.patcher.extensions.InstructionExtensions.getInstruction
import app.revanced.patcher.patch.PatchException
import app.revanced.patcher.patch.bytecodePatch
import app.revanced.patcher.patch.resourcePatch
import app.revanced.patches.all.misc.resources.addResources
import app.revanced.patches.all.misc.resources.addResourcesPatch
import app.revanced.patches.music.misc.extension.sharedExtensionPatch
import app.revanced.patches.music.misc.settings.PreferenceScreen
import app.revanced.patches.music.misc.settings.settingsPatch
import app.revanced.patches.shared.misc.mapping.get
import app.revanced.patches.shared.misc.mapping.resourceMappingPatch
import app.revanced.patches.shared.misc.mapping.resourceMappings
import app.revanced.patches.shared.misc.settings.preference.PreferenceScreenPreference
import app.revanced.patches.shared.misc.settings.preference.SwitchPreference
import app.revanced.util.indexOfFirstInstructionOrThrow
import app.revanced.util.indexOfFirstLiteralInstructionOrThrow
import com.android.tools.smali.dexlib2.Opcode
import com.android.tools.smali.dexlib2.iface.instruction.FiveRegisterInstruction
import com.android.tools.smali.dexlib2.iface.instruction.OneRegisterInstruction
import com.android.tools.smali.dexlib2.iface.instruction.ReferenceInstruction

internal var text1 = -1L
    private set

private const val EXTENSION_CLASS_DESCRIPTOR = "Lapp/revanced/extension/music/patches/NavigationBarPatch;"

@Suppress("unused")
val navigationBarPatch = bytecodePatch(
    name = "Navigation bar",
    description = "Adds options to hide navigation bar, labels and buttons."
) {
    dependsOn(
        resourceMappingPatch,
        sharedExtensionPatch,
        settingsPatch,
        addResourcesPatch,
        resourcePatch {
            execute {
                // Ensure the first ImageView has 'layout_weight' to stay properly sized
                // when the TextView is hidden.
                document("res/layout/image_with_text_tab.xml").use { document ->
                    val imageView = document.getElementsByTagName("ImageView").item(0)
                    imageView?.let {
                        if (it.attributes.getNamedItem("android:layout_weight") == null) {
                            val attr = document.createAttribute("android:layout_weight")
                            attr.value = "0.5"
                            it.attributes.setNamedItem(attr)
                        }
                    }
                }
            }
        }
    )

    compatibleWith(
        "com.google.android.apps.youtube.music"(
            "7.29.52",
            "8.10.52"
        )
    )

    execute {
        text1 = resourceMappings["id", "text1"]

        addResources("music", "layout.navigationbar.navigationBarPatch")

        PreferenceScreen.GENERAL.addPreferences(
            PreferenceScreenPreference(
                key = "revanced_music_navigation_bar_screen",
                sorting = PreferenceScreenPreference.Sorting.UNSORTED,
                preferences = setOf(
                    SwitchPreference("revanced_music_hide_navigation_bar_home_button"),
                    SwitchPreference("revanced_music_hide_navigation_bar_samples_button"),
                    SwitchPreference("revanced_music_hide_navigation_bar_explore_button"),
                    SwitchPreference("revanced_music_hide_navigation_bar_library_button"),
                    SwitchPreference("revanced_music_hide_navigation_bar_upgrade_button"),

                    SwitchPreference("revanced_music_hide_navigation_bar"),
                    SwitchPreference("revanced_music_hide_navigation_bar_labels"),
                )
            )
        )

        tabLayoutTextFingerprint.method.apply {
            // Hide navigation labels.
            val constIndex = indexOfFirstLiteralInstructionOrThrow(text1)
            val targetIndex = indexOfFirstInstructionOrThrow(constIndex, Opcode.CHECK_CAST)
            val targetParameter = getInstruction<ReferenceInstruction>(targetIndex).reference
            val targetRegister = getInstruction<OneRegisterInstruction>(targetIndex).registerA

            if (!targetParameter.toString().endsWith("Landroid/widget/TextView;"))
                throw PatchException("Method signature parameter did not match: $targetParameter")

            addInstruction(
                targetIndex + 1,
                "invoke-static { v$targetRegister }, $EXTENSION_CLASS_DESCRIPTOR->hideNavigationLabel(Landroid/widget/TextView;)V"
            )

            // Set navigation enum and hide navigation buttons.
            val enumIndex = tabLayoutTextFingerprint.patternMatch!!.startIndex + 3
            val enumRegister = getInstruction<OneRegisterInstruction>(enumIndex).registerA
            val insertEnumIndex = indexOfFirstInstructionOrThrow(Opcode.AND_INT_LIT8) - 2

            val pivotTabIndex = indexOfGetVisibilityInstruction(this)
            val pivotTabRegister = getInstruction<FiveRegisterInstruction>(pivotTabIndex).registerC

            addInstruction(
                pivotTabIndex,
                "invoke-static { v$pivotTabRegister }, $EXTENSION_CLASS_DESCRIPTOR->hideNavigationButton(Landroid/view/View;)V"
            )

            addInstruction(
                insertEnumIndex,
                "invoke-static { v$enumRegister }, $EXTENSION_CLASS_DESCRIPTOR->setLastAppNavigationEnum(Ljava/lang/Enum;)V"
            )
        }
    }
}
